### PR TITLE
[WIRES] Use wire class in operations

### PR DESCRIPTION
**Pull request by mariaschuld**
This PR is part of the wire management refactor. It changes the ``wires`` attribute of operators to be a ``Wires`` instance, but then turns it back into a list wherever the attribute is used. At this stage, we still assume that wires are sequences of consecutive integers. 

**Changes**

* Uses a ``Wires`` instance for the ``wires`` attribute in the ``Operator`` class, but converts wires back to lists everywhere the attribute is used. 

* A lot of tests of the form ``assert xyz.wires == [1]`` were updated to ``assert xyz.wires == Wires([1])``.

* A few TODO tags where dropped to mark obvious places where future PRs in this series need to check the wire ordering.

* The templates can now use ``Wires`` objects, and conversions to lists were deleted.

### CODE REVIEW

File: pennylane/operation.py [hunk 1]
```diff
@@ -364,6 +365,19 @@ def __init__(self, *params, wires=None, do_queue=True):
         if wires is None:
             raise ValueError("Must specify the wires that {} acts on".format(self.name))
 
+        # turn wires into Wires object
+        self._wires = Wires(wires)  #: pennylane.wires.Wires: wires on which the operator acts
```
Comment (mariaschuld) [inline pennylane/operation.py:369]:
```suggestion
        self._wires = Wires(wires)  #: Wires: wires on which the operator acts
```

File: pennylane/beta/plugins/default_tensor.py [hunk 2]
```diff
@@ -490,6 +490,7 @@ def expval(self, observable, wires, par):
 
         tensors = []
         for o, p, w in zip(observable, par, wires):
+
```
Comment (mariaschuld) [inline pennylane/beta/plugins/default_tensor.py:493]:
```suggestion
```

File: pennylane/operation.py [hunk 3]
```diff
@@ -364,6 +365,19 @@ def __init__(self, *params, wires=None, do_queue=True):
         if wires is None:
             raise ValueError("Must specify the wires that {} acts on".format(self.name))
 
+        # turn wires into Wires object
+        self._wires = Wires(wires)  #: Wires: wires on which the operator acts
+
+        if (
```
Comment (mariaschuld) [inline pennylane/operation.py:371]:
```suggestion
        # check that the number of wires given corresponds to required number
        if (
```

File: tests/ops/test_cv_ops.py [hunk 4]
```diff
@@ -188,15 +188,15 @@ def test_heisenberg_transformation_nongaussian(self):
         Heisenberg transformation of non-Gaussian gates."""
         op = cv.Kerr
         with pytest.raises(RuntimeError, match=r"not a Gaussian operation"):
-            op_ = op(*[0.1] * op.num_params, wires=[0] * op.num_wires)
+            op_ = op(*[0.1] * op.num_params, wires=range(op.num_wires))
```
Comment (mariaschuld) [inline tests/ops/test_cv_ops.py:191]: I'm not sure what the reason for ``[0]*op.num_wires`` was, but it is an illegal wires argument since it repeats the zero. Maybe this was just not important for the error check here, but now it is.

File: pennylane/operation.py [hunk 5]
```diff
@@ -376,18 +391,12 @@ def __init__(self, *params, wires=None, do_queue=True):
                 self.check_domain(p)
         self.params = list(params)  #: list[Any]: parameters of the operator
```
Comment (mariaschuld) [inline pennylane/operation.py:393]: These checks are now unnecessary...

File: pennylane/operation.py [hunk 6]
```diff
@@ -364,6 +365,20 @@ def __init__(self, *params, wires=None, do_queue=True):
         if wires is None:
             raise ValueError("Must specify the wires that {} acts on".format(self.name))
 
+        # turn wires into Wires object
+        self._wires = Wires(wires)  #: Wires: wires on which the operator acts
+
+        # check that the number of wires given corresponds to required number
```
Comment (mariaschuld) [inline pennylane/operation.py:371]: This check was previously in a much longer ``_check_wires`` function that is now unnecessary.

File: pennylane/operation.py [hunk 7]
```diff
@@ -229,8 +230,8 @@ class Operator(abc.ABC):
         params (tuple[float, int, array, Variable]): operator parameters
 
     Keyword Args:
-        wires (Sequence[int]): Subsystems it acts on. If not given, args[-1]
-            is interpreted as wires.
+        wires (Iterable or Wires): Iterable containing representations of the wires that the operator acts on,
+            or Wires object. If not given, args[-1] is interpreted as wires.
```
Comment (mariaschuld) [inline pennylane/operation.py:234]: Technically, this argument can be ``Iterable[Number, str] or Number or str or Wires``. I was trying forth and back to get a concise but truthful description...Shall I add more detail?

File: pennylane/templates/broadcast.py [hunk 8]
```diff
@@ -572,5 +572,4 @@ def circuit(pars):
 
     # broadcast the unitary
     for wires, pars in zip(wire_sequence[pattern], parameters):
```
Comment (mariaschuld) [inline pennylane/templates/broadcast.py:574]: What follows below are the deletions of the templates refactor. Templates can now pass their wires arguments through the operations they call.

File: pennylane/templates/subroutines/interferometer.py [hunk 9]
```diff
@@ -151,10 +149,10 @@ def Interferometer(theta, phi, varphi, wires, mesh="rectangular", beamsplitter="
                 # skip even or odd pairs depending on layer
                 if (l + k) % 2 != 1:
                     if beamsplitter == "clements":
```
Comment (mariaschuld) [inline pennylane/templates/subroutines/interferometer.py:151]: I forgot these two changes below in the templates refactor...

File: pennylane/templates/subroutines/single_excitation_unitary.py [hunk 10]
```diff
@@ -137,9 +137,7 @@ def circuit(weight, ph=None):
 
     ###############
 
-    wires = wires.tolist()  # TODO: Remove when ops accept wires
-
-    r, p = wires
+    r, p = wires.tolist()  # TODO: need to change logic here when introducing non-consec wires
```
Comment (mariaschuld) [inline pennylane/templates/subroutines/single_excitation_unitary.py:140]: This still needs to be a list, because the qchem templates heavily depend on wires containing integers which are qubit indices. Need to rewrite those...

File: tests/ops/test_qubit_ops.py [hunk 11]
```diff
@@ -18,10 +18,9 @@
 import functools
 import numpy as np
 from numpy.linalg import multi_dot
-from scipy.linalg import block_diag
 
 import pennylane as qml
-from pennylane.templates.layers import StronglyEntanglingLayers
```
Comment (mariaschuld) [inline tests/ops/test_qubit_ops.py:24]: That one was never used...

File: tests/templates/test_subroutines.py [hunk 12]
```diff
@@ -804,7 +805,7 @@ def circuit(weights=weights, wires=wires, ph=ph, pphh=pphh, init_state=init_stat
         ("weights", "ph", "pphh", "expected"),
         [
             (np.array([3.90575761, -1.89772083, -1.36689032]),
-             [[0, 2], [1, 3]], [0, 1, 2, 3],
+             [[0, 2], [1, 3]], [[0, 1, 2, 3]],
```
Comment (mariaschuld) [inline tests/templates/test_subroutines.py:808]: The bugfix in the test uncovered a bug in the fixtures.

File: tests/test_operation.py [hunk 13]
```diff
@@ -294,7 +295,7 @@ class DummyOp(qml.operation.Operator):
             num_params = 1
             par_domain = 'R'
 
-        with pytest.raises(ValueError, match="wires must be unique"):
+        with pytest.raises(qml.wires.WireError, match="Wires must be unique"):
```
Comment (mariaschuld) [inline tests/test_operation.py:298]: Didn't want to import for only one line...

**Comment by mariaschuld** [comment]
Please ignore the coverage for now, I will fix this before merging.

File: pennylane/operation.py [hunk 14]
```diff
@@ -396,41 +405,8 @@ def __repr__(self):
...
-        ):
-            raise ValueError(
-                "{}: wrong number of wires. "
-                "{} wires given, {} expected.".format(self.name, len(wires), self.num_wires)
-            )
-
-        if len(set(wires)) != len(wires):
-            raise ValueError("{}: wires must be unique, got {}.".format(self.name, wires))
-
-        return tuple(int(w) for w in wires)
```
Comment (josh146) [inline pennylane/operation.py:433]: awesome that all this can be removed :)

File: pennylane/operation.py [hunk 15]
```diff
@@ -1282,7 +1259,7 @@ def loc(w):
         if U.ndim == 1:
             W = np.zeros(dim)
             W[0] = U[0]
-            for k, w in enumerate(self.wires):
+            for k, w in enumerate(self.wires.tolist()):
```
Comment (josh146) [inline pennylane/operation.py:1262]: Same question here

File: pennylane/operation.py [hunk 16]
```diff
@@ -1292,7 +1269,7 @@ def loc(w):
 
             W[0, 0] = U[0, 0]
 
-            for k1, w1 in enumerate(self.wires):
+            for k1, w1 in enumerate(self.wires.tolist()):
```
Comment (josh146) [inline pennylane/operation.py:1272]: Same question here

File: pennylane/operation.py [hunk 17]
```diff
@@ -1301,7 +1278,7 @@ def loc(w):
                 # first row (for gates, the first row is always (1, 0, 0, ...), but not for observables!)
                 W[0, d1] = U[0, s1]
 
-                for k2, w2 in enumerate(self.wires):
+                for k2, w2 in enumerate(self.wires.tolist()):
```
Comment (josh146) [inline pennylane/operation.py:1281]: Same question here

File: pennylane/plugins/default_gaussian.py [hunk 18]
```diff
@@ -841,7 +841,7 @@ def reduced_state(self, wires):
         r""" Returns the vector of means and the covariance matrix of the specified wires.
 
         Args:
-            wires (int of Sequence[int]): indices of the requested wires
+            wires (Iterable[int] or int): indices of the requested wires # TODO: re-assess for nonconsec wires
```
Comment (josh146) [inline pennylane/plugins/default_gaussian.py:844]: Probably best not to have the comment in the docstring 😆

File: pennylane/qnodes/base.py [hunk 19]
```diff
@@ -305,14 +305,14 @@ def _remove_operator(self, operator):
...
-            if set(operator.wires) != set(range(self.num_wires)):
+        if operator.num_wires == ActsOn.AllWires:  # TODO: re-assess for nonconsec wires
+            if set(operator.wires.tolist()) != set(range(self.num_wires)):
                 raise QuantumFunctionError(
                     "Operator {} must act on all wires".format(operator.name)
                 )
 
         # Make sure only existing wires are used.
-        for w in _flatten(operator.wires):
+        for w in operator.wires.tolist():  # TODO: re-assess for for nonconsec wires
```
Comment (josh146) [inline pennylane/qnodes/base.py:315]: Same question here

File: pennylane/qnodes/base.py [hunk 20]
```diff
@@ -305,14 +305,14 @@ def _remove_operator(self, operator):
             self.queue.remove(operator)
 
     def _append_operator(self, operator):
-        if operator.num_wires == ActsOn.AllWires:
-            if set(operator.wires) != set(range(self.num_wires)):
+        if operator.num_wires == ActsOn.AllWires:  # TODO: re-assess for nonconsec wires
+            if set(operator.wires.tolist()) != set(range(self.num_wires)):
```
Comment (josh146) [inline pennylane/qnodes/base.py:309]: Same question here

File: tests/ops/test_qubit_ops.py (hunk 11 above)
Comment (josh146) [reply tests/ops/test_qubit_ops.py:24]: 😆

File: tests/qnodes/test_qnode_base.py [hunk 21]
```diff
@@ -521,11 +522,11 @@ def test_bad_wire_argument(self, operable_mock_device_2_wires):
         """Error: wire arguments must be intergers."""
 
         def circuit(x):
-            qml.RX(x, wires=[0.5])
+            qml.RX(x, wires=[qml.PauliX])
```
Comment (josh146) [inline tests/qnodes/test_qnode_base.py:525]: 😆 

(Out of curiosity, what was the reason for changing it from 0.5? Shouldn't that also raise an error?)

File: pennylane/_qubit_device.py [hunk 22]
```diff
@@ -195,7 +195,7 @@ def apply(self, operations, **kwargs):
         >>> op = qml.RX(0.2, wires=[0])
         >>> op.name # returns the operation name
         "RX"
-        >>> op.wires # returns a list of wires
+        >>> op.wires.tolist() # returns a list of wires
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/_qubit_device.py:198]: Ok, true. I thought we'd tell developers how to get the list out of the object, but you are right!

File: pennylane/_qubit_device.py [hunk 23]
```diff
@@ -382,7 +382,7 @@ def estimate_probability(self, wires=None):
         # consider only the requested wires
         wires = np.hstack(wires)
 
-        samples = self._samples[:, np.array(wires)]
+        samples = self._samples[:, np.array(wires)]  # TODO: Use indices for nonconsec wires
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/_qubit_device.py:385]: If ``samples`` is a list of 3 elements, but the wires is a sequence like [5, 10] (maybe ``5`` addressing the first qubit and ``10`` the third), then we cannot index the former by the latter. Instead, my plan is to turn the user-given ``op.wires`` into a list of indices of the wires in the device's register (like ``[0, 2]`` in this case). Then all devices should work again. 

I'm just flagging the locations where I spotted that this will be necessary. The implementation will be done in the next PR.

File: pennylane/_qubit_device.py [hunk 24]
```diff
@@ -227,7 +227,7 @@ def active_wires(operators):
         """
         wires = []
         for op in operators:
-            wires.extend(op.wires)
+            wires.extend(op.wires.tolist())
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/_qubit_device.py:230]: Ah, you are right. This might pop up elsewhere, it was from a stage where the Wires class' ``extend`` method took lists. Thanks!

File: pennylane/circuit_drawer/circuit_drawer.py [hunk 25]
```diff
@@ -231,9 +231,9 @@ def resolve_decorations(self, grid, representation_grid):
 
                 if isinstance(op, qml.operation.Tensor):
                     # pylint: disable=protected-access
-                    wires = list(qml.utils._flatten(op.wires))
+                    wires = list(qml.utils._flatten(op.wires.tolist()))
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/circuit_drawer/circuit_drawer.py:234]: Yes, I think so. Even if the ``_flatten`` function does not complain, I don't think you can parse Wires to list. 

But after PR #656 we can actually simplify this line to ``wires = op.wires.tolist()``.

File: pennylane/_qubit_device.py [hunk 26]
```diff
@@ -495,7 +495,7 @@ def var(self, observable):
         return np.var(self.sample(observable))
 
     def sample(self, observable):
-        wires = observable.wires
+        wires = observable.wires.tolist()  # TODO: re-asses for nonconsecutive wires
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/_qubit_device.py:498]: The sample function also assumes that ``wires`` is a subset from a list of consecutive integers. If we allow for non-consecutive wires this needs to be re-assessed. Just flagging it here.

File: pennylane/circuit_drawer/circuit_drawer.py [hunk 27]
```diff
@@ -310,8 +310,8 @@ def move_multi_wire_gates(operator_grid):
                 if op is None:
                     continue
 
-                if len(op.wires) > 1:
-                    sorted_wires = op.wires.copy()
+                if len(op.wires.tolist()) > 1:
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/circuit_drawer/circuit_drawer.py:313]: You got me here. Good point!

File: pennylane/circuit_graph.py [hunk 28]
```diff
@@ -155,10 +155,10 @@ def __init__(self, ops, variable_deps):
         self.num_wires = 0
         """int: number of wires the circuit contains"""
         for k, op in enumerate(ops):
-            self.num_wires = max(self.num_wires, max(list(_flatten(op.wires))) + 1)
+            self.num_wires = max(self.num_wires, max(list(_flatten(op.wires.tolist()))) + 1)
             op.queue_idx = k  # store the queue index in the Operator
             for w in set(
-                _flatten(op.wires)
+                _flatten(op.wires.tolist())
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/circuit_graph.py:161]: Again the case as above. In principle yes, but after #656 we can remove the ``_flatten`` function. I should have done that in that PR, thinking about it!

[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/circuit_graph.py:161]: Even then the ``tolist()`` is needed, because the iteration will pick ``w`` which are ``Wires`` of length one (remember, the ``__it__`` function returns Wire objects). And then the ``_grid`` in the next line will be made up of Wires objects and all hell breaks loose.

File: pennylane/circuit_graph.py [hunk 29]
```diff
@@ -155,10 +155,10 @@ def __init__(self, ops, variable_deps):
         self.num_wires = 0
         """int: number of wires the circuit contains"""
         for k, op in enumerate(ops):
-            self.num_wires = max(self.num_wires, max(list(_flatten(op.wires))) + 1)
+            self.num_wires = max(self.num_wires, max(list(_flatten(op.wires.tolist()))) + 1)
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/circuit_graph.py:158]: Again, ``_flatten`` is superfluous, but ``tolist()`` is not. 

This line will change again in the next PR, which assumes a user can work with a register of non-consecutive or even string-named wires (which does not define a max function). I need to think about how to do this elegantly here.

File: pennylane/circuit_graph.py [hunk 30]
```diff
@@ -307,7 +307,7 @@ class QASMSerializerDevice:
         # create the QASM code representing the operations
         for op in decomposed_ops:
             gate = OPENQASM_GATES[op.name]
-            wires = ",".join(["q[{}]".format(w) for w in op.wires])
+            wires = ",".join(["q[{}]".format(w) for w in op.wires.tolist()])
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/circuit_graph.py:310]: Again necessary, because we don't want ``format`` to print an object like `` "<Wires=[0]>" ``.

File: pennylane/plugins/default_gaussian.py (hunk 18 above)
Comment (mariaschuld) [reply pennylane/plugins/default_gaussian.py:844]: You are right. I deleted it altogether, I will remember this tag anyways.

File: tests/qnodes/test_qnode_base.py (hunk 21 above)
Comment (mariaschuld) [reply tests/qnodes/test_qnode_base.py:525]: Nope. At the moment the user can choose pretty much any number or string as a wire. We can change this if it's too loose. But if you think about it, we should not care what they call their wires, as long as we can identify the index at which this object sits in the device's record of all wires. PennyLane will only really work with the indices.

**Comment by mariaschuld** [comment]
@josh146 Thanks for your review, you found a couple of things I overlooked. 

As mentioned, I think most of your comments may be resolved by noting that an iteration over a Wires object returns Wires objects of length one:

File: tests/qnodes/test_qnode_base.py (hunk 21 above)
Comment (mariaschuld) [reply tests/qnodes/test_qnode_base.py:525]: Maybe we should constrain Wires from Numbers to Integral type?

File: pennylane/_qubit_device.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_qubit_device.py]: Ah, we could include a link here to the `Wires` class directly

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_qubit_device.py]: That way all the info about the wires class is in one place (likely we would forget to update it here if the wires class changes in the future)

File: pennylane/_qubit_device.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_qubit_device.py:497]: Sorry for posting the same comment in so many places 😆

File: tests/qnodes/test_qnode_base.py (hunk 21 above)
Comment (josh146) [reply tests/qnodes/test_qnode_base.py:525]: I wonder if there are odd edge cases... will the class treat `1.` and `1` as two separate wires?

Comment (mariaschuld) [reply tests/qnodes/test_qnode_base.py:525]: YOu are right. I will restrict the objects that can represent wires to Integrals in a separate PR!

**Comment by mariaschuld** [comment]
@co9olguy I wonder if you could add your review? The main PR to remove all the ``tolist()`` is yet to come :)

File: pennylane/_qubit_device.py [hunk 31]
```diff
@@ -214,7 +215,7 @@ def apply(self, operations, **kwargs):
         >>> op = qml.RX(0.2, wires=[0])
         >>> op.name # returns the operation name
         "RX"
-        >>> op.wires # returns a list of wires
+        >>> op.wires # returns a Wires object representing the wires that the operation acts on
         [0]
```
Comment (co9olguy) [inline pennylane/_qubit_device.py:219]:
```suggestion
        Wires([0])
```
?

File: pennylane/circuit_graph.py [hunk 32]
```diff
@@ -155,11 +155,9 @@ def __init__(self, ops, variable_deps):
         self.num_wires = 0
         """int: number of wires the circuit contains"""
         for k, op in enumerate(ops):
-            self.num_wires = max(self.num_wires, max(list(_flatten(op.wires))) + 1)
+            self.num_wires = max(self.num_wires, max(op.wires.tolist()) + 1)
```
Comment (co9olguy) [inline pennylane/circuit_graph.py:158]: @mariaschuld might be worth putting a todo here. Looks like this logic is implicitly assuming the last index is related to the number of wires

File: pennylane/operation.py [hunk 33]
```diff
@@ -364,6 +365,20 @@ def __init__(self, *params, wires=None, do_queue=True):
...
+
+        # check that the number of wires given corresponds to required number
+        if (
+            self.num_wires != AllWires
+            and self.num_wires != AnyWires
+            and len(self._wires) != self.num_wires
+        ):
+            raise ValueError(
+                "{}: wrong number of wires. "
+                "{} wires given, {} expected.".format(self.name, len(wires), self.num_wires)
```
Comment (co9olguy) [inline pennylane/operation.py:379]: I think we want to avoid using `len(wires)` here, since a string would give a confusing error message:
```
qml.RX(0.5, wires="qubit0")  # len(wires)==6
```
```suggestion
                "{} wires given, {} expected.".format(self.name, len(self._wires), self.num_wires)
```

File: pennylane/plugins/default_gaussian.py [hunk 34]
```diff
@@ -841,7 +841,7 @@ def reduced_state(self, wires):
         r""" Returns the vector of means and the covariance matrix of the specified wires.
 
         Args:
-            wires (int of Sequence[int]): indices of the requested wires
+            wires (Iterable[int] or int): indices of the requested wires
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py:844]: so `reduced_state(["mode5", "mode8"])` wouldn't be allowed?

File: pennylane/plugins/default_qubit.py [hunk 35]
```diff
@@ -121,23 +121,25 @@ def _apply_operation(self, operation):
         Args:
             operation (~.Operation): operation to apply on the device
         """
+        wires = operation.wires.tolist()  # TODO: translation to nonconsec wires indices
+
         if isinstance(operation, QubitStateVector):
-            self._apply_state_vector(operation.parameters[0], operation.wires)
```
Comment (co9olguy) [inline pennylane/plugins/default_qubit.py:125]: default.qubit was very recently refactored. hopefully we don't hit any merge conflicts

File: tests/circuit_drawer/test_circuit_drawer.py [hunk 36]
```diff
@@ -93,7 +93,7 @@ def to_layer(operation_list, num_wires):
     layer = [None] * num_wires
```
Comment (co9olguy) [inline tests/circuit_drawer/test_circuit_drawer.py:93]: Just commenting here but really I have a question about another file:
@josh146 is the file `qchem/tmp1v5rrgvd` something that is needed in version control?

Comment (josh146) [reply tests/circuit_drawer/test_circuit_drawer.py:93]: Nope, wonder where it came from 🤔

File: pennylane/circuit_graph.py (hunk 32 above)
Comment (mariaschuld) [reply pennylane/circuit_graph.py:158]: Yes, I have it on my radar. The ``tolist()`` serve as implicit ``TODO``s...

File: pennylane/plugins/default_qubit.py (hunk 35 above)
Comment (mariaschuld) [reply pennylane/plugins/default_qubit.py:125]: So far all clean!

File: pennylane/plugins/default_gaussian.py (hunk 34 above)
Comment (mariaschuld) [reply pennylane/plugins/default_gaussian.py:844]: Good point. Need to go over default_gaussian again to adapt the docstrings.
